File: source/includes/sample_apps/csfle/java-fle-2/src/main/java/com/mongodb/csfle/insertEncryptedDocument.java
package com.mongodb.csfle;
/*
 * Copyright 2008-present MongoDB, Inc.

 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */


import java.util.*;

import java.util.HashMap;
import java.util.Map;

import static com.mongodb.client.model.Filters.eq;

import com.mongodb.AutoEncryptionSettings;
import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoClients;
import com.mongodb.client.MongoCollection;

import org.bson.BsonArray;
import org.bson.BsonBinary;
import org.bson.BsonDocument;
import org.bson.BsonString;
import org.bson.types.Binary;
import org.bson.BsonBinarySubType;

import com.mongodb.ConnectionString;
import com.mongodb.MongoClientSettings;
import org.bson.Document;

// :state-start: local-reader local-test
import java.io.FileInputStream;
// :state-end:


/*
 * - Reads master key from file "master-key.txt" in root directory of project
 * - Creates a JSON schema for a specified collection to enable automatic encryption
 * - Creates an encrypted client and upserts a single document
 * - Finds the upserted document with the encrypted client using an encrypted field
 * - Attempts to find the upserted document with the normal client using an encrypted field
 * - Finds the upserted document with the normal client using a non-encrypted field
 */
public class insertEncryptedDocument {

    public static void main(String[] args) throws Exception {


        String encryptedDbName = "medicalRecords";
        String encryptedCollName = "patients";
        String encryptedNameSpace = encryptedDbName + "." + encryptedCollName;

        // start-key-vault
        String keyVaultDb = "encryption";
        String keyVaultColl = "__keyVault";
        String keyVaultNamespace = keyVaultDb + "." + keyVaultColl;
        // end-key-vault

        // :state-start: aws-test azure-test local-test gcp-test
        String connectionString = System.getenv("MONGODB_URI");
        // :state-end:
        // :state-start: local-reader aws-reader azure-reader gcp-reader
        // :uncomment-start:
        //String connectionString = "<Your MongoDB URI>";
        // :uncomment-end:
        // :state-end:

        // start-kmsproviders
        // :state-start: local-reader local-test
        String kmsProvider = "local";
        String path = "master-key.txt";

        byte[] localMasterKeyRead = new byte[96];

        try (FileInputStream fis = new FileInputStream(path)) {
            if (fis.read(localMasterKeyRead) < 96)
                throw new Exception("Expected to read 96 bytes from file");
        }
        Map<String, Object> keyMap = new HashMap<String, Object>();
        keyMap.put("key", localMasterKeyRead);

        Map<String, Map<String, Object>> kmsProviders = new HashMap<String, Map<String, Object>>();
        kmsProviders.put(kmsProvider, keyMap);
        // :state-end:
        // :state-uncomment-start: aws-reader
        //Map<String, Map<String, Object>> kmsProviders = new HashMap<String, Map<String, Object>>();
        //String kmsProvider = "aws";
        //Map<String, Object> providerDetails = new HashMap<>();
        //providerDetails.put("accessKeyId", "<IAM User Access Key ID>");
        //providerDetails.put("secretAccessKey", "<IAM User Secret Access Key>");
        //kmsProviders.put(kmsProvider, providerDetails);
        // :state-uncomment-end:
        // :state-uncomment-start: aws-test
        //Map<String, Map<String, Object>> kmsProviders = new HashMap<String, Map<String, Object>>();
        //String kmsProvider = "aws";
        //Map<String, Object> providerDetails = new HashMap<>();
        //providerDetails.put("accessKeyId", System.getenv("AWS_ACCESS_KEY_ID"));
        //providerDetails.put("secretAccessKey", System.getenv("AWS_SECRET_ACCESS_KEY"));
        //kmsProviders.put(kmsProvider, providerDetails);
        // :state-uncomment-end:
        // :state-uncomment-start: azure-reader
        //String kmsProvider = "azure";
        //Map<String, Map<String, Object>> kmsProviders = new HashMap<String, Map<String, Object>>();
        //Map<String, Object> providerDetails = new HashMap<>();
        //providerDetails.put("tenantId", "<Azure account organization>");
        //providerDetails.put("clientId", "<Azure client ID>");
        //providerDetails.put("clientSecret", "<Azure client secret>");
        //kmsProviders.put(kmsProvider, providerDetails);
        // :state-uncomment-end:
        // :state-uncomment-start: azure-test
        //Map<String, Map<String, Object>> kmsProviders = new HashMap<String, Map<String, Object>>();
        //String kmsProvider = "azure";
        //Map<String, Object> providerDetails = new HashMap<>();
        //providerDetails.put("tenantId", System.getenv("AZURE_TENANT_ID"));
        //providerDetails.put("clientId", System.getenv("AZURE_CLIENT_ID"));
        //providerDetails.put("clientSecret", System.getenv("AZURE_CLIENT_SECRET"));
        //kmsProviders.put(kmsProvider, providerDetails);
        // :state-uncomment-end:
        // :state-uncomment-start: gcp-reader
        //String kmsProvider = "gcp";
        //Map<String, Map<String, Object>> kmsProviders = new HashMap<String, Map<String, Object>>();
        //Map<String, Object> providerDetails = new HashMap<>();
        //providerDetails.put("email", "<Your GCP Email Address>");
        //providerDetails.put("privateKey", "<Your GCP Private Key>");
        //kmsProviders.put(kmsProvider, providerDetails);
        // :state-uncomment-end:
        // :state-uncomment-start: gcp-test
        //String kmsProvider = "gcp";
        //Map<String, Map<String, Object>> kmsProviders = new HashMap<String, Map<String, Object>>();
        //Map<String, Object> providerDetails = new HashMap<>();
        //providerDetails.put("email", System.getenv("GCP_EMAIL"));
        //providerDetails.put("privateKey", System.getenv("GCP_PRIVATE_KEY"));
        //kmsProviders.put(kmsProvider, providerDetails);
        // :state-uncomment-end:
        // end-kmsproviders

        // start-schema
        MongoClient regularClient = MongoClients.create(connectionString);

        MongoCollection<Document> keyVaultClient = regularClient.getDatabase(keyVaultDb).getCollection(keyVaultColl);

        BsonBinary dataKeyId1 = new BsonBinary(BsonBinarySubType.UUID_STANDARD, keyVaultClient.find(eq("keyAltNames", "dataKey1")).first().get("_id", Binary.class).getData());
        BsonBinary dataKeyId2 = new BsonBinary(BsonBinarySubType.UUID_STANDARD, keyVaultClient.find(eq("keyAltNames", "dataKey2")).first().get("_id", Binary.class).getData());
        BsonBinary dataKeyId3 = new BsonBinary(BsonBinarySubType.UUID_STANDARD, keyVaultClient.find(eq("keyAltNames", "dataKey3")).first().get("_id", Binary.class).getData());
        BsonBinary dataKeyId4 = new BsonBinary(BsonBinarySubType.UUID_STANDARD, keyVaultClient.find(eq("keyAltNames", "dataKey4")).first().get("_id", Binary.class).getData());

        BsonDocument encFields = new BsonDocument().append("fields",
                new BsonArray(Arrays.asList(
                        new BsonDocument().append("keyId", dataKeyId1)
                                .append("path", new BsonString("patientId"))
                                .append("bsonType", new BsonString("int"))
                                .append("queries", new BsonDocument().append("queryType", new BsonString("equality"))),
                        new BsonDocument().append("keyId", dataKeyId2)
                                .append("path", new BsonString("medications"))
                                .append("bsonType", new BsonString("array")),
                        new BsonDocument().append("keyId", dataKeyId3)
                                .append("path", new BsonString("patientRecord.ssn"))
                                .append("bsonType", new BsonString("string"))
                                .append("queries", new BsonDocument().append("queryType", new BsonString("equality"))),
                        new BsonDocument().append("keyId", dataKeyId4)
                                .append("path", new BsonString("patientRecord.billing"))
                                .append("bsonType", new BsonString("object")))));
        Map<String, BsonDocument> encryptedFieldsMap = new HashMap<String, BsonDocument>();
        encryptedFieldsMap.put(encryptedNameSpace, encFields);

        // end-schema

        // start-extra-options
        Map<String, Object> extraOptions = new HashMap<String, Object>();
        // :state-start: aws-reader azure-reader local-reader gcp-reader
        // :uncomment-start:
        //extraOptions.put("cryptSharedLibPath", "<path to crypt_shared>");
        // :uncomment-end:
        // :state-end:
        // :state-start: aws-test azure-test local-test gcp-test
        extraOptions.put("cryptSharedLibPath", System.getenv("SHARED_LIB_PATH"));
        // :state-end:
        // end-extra-options

        // start-client
        MongoClientSettings clientSettings = MongoClientSettings.builder()
                .applyConnectionString(new ConnectionString(connectionString))
                .autoEncryptionSettings(AutoEncryptionSettings.builder()
                        .keyVaultNamespace(keyVaultNamespace)
                        .kmsProviders(kmsProviders)
                        .encryptedFieldsMap(encryptedFieldsMap)
                        .extraOptions(extraOptions)
                        .build())
                .build();
        MongoClient mongoClientSecure = MongoClients.create(clientSettings);
        // end-client


        // start-insert
        ArrayList<String> medications = new ArrayList<>();
        medications.add("Atorvastatin");
        medications.add("Levothyroxine");

        Document patientRecord = new Document()
                .append("ssn", "987-65-4320")
                .append("billing", new Document().append("type", "Visa").append("number", "4111111111111111"));

        Document patient = new Document()
                .append("firstName", "Jon")
                .append("lastName", "Doe")
                .append("patientId", 12345678)
                .append("address", "AB+")
                .append("medications", medications)
                .append("patientRecord", patientRecord);
        mongoClientSecure.getDatabase(encryptedDbName).getCollection(encryptedCollName).insertOne(patient);
        // end-insert

        // start-find
        System.out.println("Finding a document with regular (non-encrypted) client.");
        Document docRegular = regularClient.getDatabase(encryptedDbName).getCollection(encryptedCollName).find(eq("firstName", "Jon")).first();
        System.out.println(docRegular.toJson());
        System.out.println("Finding a document with encrypted client, searching on an encrypted field");
        Document docSecure = mongoClientSecure.getDatabase(encryptedDbName).getCollection(encryptedCollName).find(eq("firstName", "Jon")).first();
        System.out.println(docSecure.toJson());
        // end-find 

        mongoClientSecure.close();
        regularClient.close();

    }
}

File: source/includes/sample_apps/csfle/dotnet-fle-2/QueryableEncryption/InsertEncryptedDocument.cs
using System;
// :state-start: local-test local-reader
using System.IO;
// :state-end:
using System.Collections.Generic;
using MongoDB.Driver;
using MongoDB.Bson;
using MongoDB.Driver.Encryption;

namespace QueryableEncryption
{
    internal static class InsertEncryptedDocument
    {
        public static void Insert()
        {
            // :state-start: local-reader azure-reader aws-reader gcp-reader
            // :uncomment-start:
            //var connectionString = "<Your MongoDB URI>";
            // :uncomment-end:
            // :state-end:
            // :state-start: local-test azure-test aws-test gcp-test
            var connectionString = Environment.GetEnvironmentVariable("MONGODB_URI");
            // :state-end:
            // start-key-vault
            var keyVaultNamespace = CollectionNamespace.FromFullName("encryption.__keyVault");
            // end-key-vault

            // start-kmsproviders
            var kmsProviders = new Dictionary<string, IReadOnlyDictionary<string, object>>();
            // :state-start: local-reader local-test
            const string provider = "local";
            const string localMasterKeyPath = "master-key.txt";
            var localMasterKeyBase64Read = File.ReadAllText(localMasterKeyPath);
            var localMasterKeyBytes = Convert.FromBase64String(localMasterKeyBase64Read);
            var localOptions = new Dictionary<string, object>
            {
                {"key", localMasterKeyBytes}
            };
            kmsProviders.Add(provider, localOptions);
            // :state-end:
            // :state-uncomment-start: aws-reader
            //const string provider = "aws";
            //var awsKmsOptions = new Dictionary<string, object>
            //{
            //    { "accessKeyId", "<Your AWS Access Key ID>" },
            //    { "secretAccessKey", "<Your AWS Secret Access Key>" }
            //};
            //kmsProviders.Add(provider, awsKmsOptions);
            // :state-uncomment-end:
            // :state-uncomment-start: aws-test
            //const string provider = "aws";
            //var awsAccessKey = Environment.GetEnvironmentVariable("AWS_ACCESS_KEY_ID");
            //var awsSecretAccessKey = Environment.GetEnvironmentVariable("AWS_SECRET_ACCESS_KEY");
            //var awsKmsOptions = new Dictionary<string, object>
            //{
            //    { "accessKeyId", awsAccessKey },
            //    { "secretAccessKey", awsSecretAccessKey }
            //};
            //kmsProviders.Add(provider, awsKmsOptions);
            // :state-uncomment-end:
            // :state-uncomment-start: azure-reader
            //const string provider = "azure";
            //var azureKmsOptions = new Dictionary<string, object>
            //{
            //    { "tenantId", "<Your Azure Tenant ID>" },
            //    { "clientId", "<Your Azure Client ID>" },
            //    { "clientSecret", "<Your Azure Client Secret>" },
            //};
            //kmsProviders.Add(provider, azureKmsOptions);
            // :state-uncomment-end:
            // :state-uncomment-start: azure-test
            //const string provider = "azure";
            //var azureTenantId = Environment.GetEnvironmentVariable("AZURE_TENANT_ID");
            //var azureClientId = Environment.GetEnvironmentVariable("AZURE_CLIENT_ID");
            //var azureClientSecret = Environment.GetEnvironmentVariable("AZURE_CLIENT_SECRET");
            //var azureKmsOptions = new Dictionary<string, object>
            //{
            //    { "tenantId", azureTenantId },
            //    { "clientId", azureClientId },
            //    { "clientSecret", azureClientSecret },
            //};
            //kmsProviders.Add(provider, azureKmsOptions);
            // :state-uncomment-end:
            // :state-uncomment-start: gcp-reader
            //const string provider = "gcp";
            //var gcpKmsOptions = new Dictionary<string, object>
            //{
            //    { "privateKey", "<Your GCP Private Key>" },
            //    { "email", "<Your GCP Email>" },
            //};
            //kmsProviders.Add(provider, gcpKmsOptions);
            // :state-uncomment-end:
            // :state-uncomment-start: gcp-test
            //const string provider = "gcp";
            //var gcpPrivateKey = Environment.GetEnvironmentVariable("GCP_PRIVATE_KEY");
            //var gcpEmail = Environment.GetEnvironmentVariable("GCP_EMAIL");
            //var gcpKmsOptions = new Dictionary<string, object>
            //{
            //    { "privateKey", gcpPrivateKey },
            //    { "email", gcpEmail },
            //};
            //kmsProviders.Add(provider, gcpKmsOptions);
            // :state-uncomment-end:
            // end-kmsproviders

            // start-schema
            var regularClient = new MongoClient(connectionString);
            var keyVaultCollection = regularClient.GetDatabase(keyVaultNamespace.DatabaseNamespace.DatabaseName)
                .GetCollection<BsonDocument>(keyVaultNamespace.CollectionName);

            BsonBinaryData GetKeyId(string altName)
            {
                var filter = Builders<BsonDocument>.Filter.Eq<BsonString>("keyAltNames", altName);
                return keyVaultCollection.Find(filter).First<BsonDocument>()["_id"].AsBsonBinaryData;
            }

            var dataKeyId1 = GetKeyId("dataKey1");
            var dataKeyId2 = GetKeyId("dataKey2");
            var dataKeyId3 = GetKeyId("dataKey3");
            var dataKeyId4 = GetKeyId("dataKey4");

            var encryptedCollectionNamespace = CollectionNamespace.FromFullName("medicalRecords.patients");
            var encryptedFieldsMap = new Dictionary<string, BsonDocument>
            {
                {
                    encryptedCollectionNamespace.FullName, new BsonDocument
                    {
                        {
                            "fields", new BsonArray
                            {
                                new BsonDocument
                                {
                                    {"keyId", dataKeyId1},
                                    {"path", new BsonString("patientId")},
                                    {"bsonType", new BsonString("int")},
                                    {
                                        "queries", new BsonDocument
                                        {
                                            {"queryType", new BsonString("equality")}
                                        }
                                    }
                                },
                                new BsonDocument
                                {
                                    {"keyId", dataKeyId2},
                                    {"path", new BsonString("medications")},
                                    {"bsonType", new BsonString("array")},
                                },
                                new BsonDocument
                                {
                                    {"keyId", dataKeyId3},
                                    {"path", new BsonString("patientRecord.ssn")},
                                    {"bsonType", new BsonString("string")},
                                    {
                                        "queries", new BsonDocument
                                        {
                                            {"queryType", new BsonString("equality")}
                                        }
                                    }
                                },
                                new BsonDocument
                                {
                                    {"keyId", dataKeyId4},
                                    {"path", new BsonString("patientRecord.billing")},
                                    {"bsonType", new BsonString("object")},
                                },
                            }
                        }
                    }
                }
            };
            // end-schema

            // start-extra-options
            // :state-start: aws-reader azure-reader local-reader gcp-reader
            // :uncomment-start:
            //var extraOptions = new Dictionary<string, object>()
            //{
            //    { "cryptSharedLibPath", "<path to crypt_shared library>" },
            //};
            // :uncomment-end:
            // :state-end:
            // :state-start: aws-test azure-test local-test gcp-test
            var extraOptions = new Dictionary<string, object>()
            {
                {"cryptSharedLibPath", Environment.GetEnvironmentVariable("SHARED_LIB_PATH")},
            };
            // :state-end:
            // end-extra-options

            // start-client
            var clientSettings = MongoClientSettings.FromConnectionString(connectionString);
            var autoEncryptionOptions = new AutoEncryptionOptions(
                keyVaultNamespace,
                kmsProviders,
                encryptedFieldsMap: encryptedFieldsMap,
                extraOptions: extraOptions);
            clientSettings.AutoEncryptionOptions = autoEncryptionOptions;
            var secureClient = new MongoClient(clientSettings);
            // end-client

            // start-insert
            var sampleDocument = new BsonDocument
            {
                {"firstName", "Jon"},
                {"lastName", "Doe"},
                {"patientId", 12345678},
                {"address", "157 Electric Ave."},
                {
                    "medications", new BsonArray
                    {
                        new BsonString("Atorvastatin"),
                        new BsonString("Levothyroxine")
                    }
                },
                {
                    "patientRecord", new BsonDocument
                    {
                        {"ssn", new BsonString("987-65-4320")},
                        {
                            "billing", new BsonDocument
                            {
                                {"type", new BsonString("Visa")},
                                {"number", "4111111111111111"}
                            }
                        }
                    }
                }
            };

            var secureCollection = secureClient.GetDatabase(encryptedCollectionNamespace.DatabaseNamespace.DatabaseName)
                .GetCollection<BsonDocument>(encryptedCollectionNamespace.CollectionName);
            secureCollection.InsertOne(sampleDocument);
            // end-insert

            // start-find
            Console.WriteLine("Finding a document with regular (non-encrypted) client.");
            var filter = Builders<BsonDocument>.Filter.Eq("firstName", "Jon");
            var regularClientEncryptedCollection = regularClient.GetDatabase(encryptedCollectionNamespace.DatabaseNamespace.DatabaseName)
                .GetCollection<BsonDocument>(encryptedCollectionNamespace.CollectionName);
            var regularResult = regularClientEncryptedCollection.Find(filter).First();
            Console.WriteLine($"\n{regularResult}\n");
            Console.WriteLine("Finding a document with encrypted client, searching on an encrypted field");
            var encryptedFieldFilter = Builders<BsonDocument>.Filter.Eq("patientRecord.ssn", "987-65-4320");
            var secureResult = secureCollection.Find(encryptedFieldFilter).First();
            Console.WriteLine($"\n{secureResult}\n");
            // end-find
        }
    }
}
